crates/libs/core/src/event.rs

```diff
@@ -25,228 +23,78 @@ impl<T: Interface> Event<T> {
     /// Creates a new, empty `Event<T>`.
     pub fn new() -> Self {
         Self {
-            delegates: Array::new(),
-            swap: Mutex::default(),
-            change: Mutex::default(),
+            delegates: RwLock::new(None),
         }
     }
 
     /// Registers a delegate with the event object.
-    pub fn add(&mut self, delegate: &T) -> Result<i64> {
-        let mut _lock_free_drop = Array::new();
-        Ok({
-            let _change_lock = self.change.lock().unwrap();
-            let mut new_delegates = Array::with_capacity(self.delegates.len() + 1)?;
-            for delegate in self.delegates.as_slice() {
-                new_delegates.push(delegate.clone());
-            }
-            let delegate = Delegate::new(delegate)?;
-            let token = delegate.to_token();
-            new_delegates.push(delegate);
+    pub fn add(&self, delegate: &T) -> Result<i64> {
+        let mut guard = self.delegates.write().unwrap();
+        let new_delegate = Delegate::new(delegate)?;
+        let token = new_delegate.to_token();
+        let new_iter = once(new_delegate);
+
+        let new_list = if let Some(old_delegates) = guard.as_ref() {
+            Arc::from_iter(old_delegates.iter().cloned().chain(new_iter))
+        } else {
+            Arc::from_iter(new_iter)
+        };
 
-            let _swap_lock = self.swap.lock().unwrap();
-            _lock_free_drop = self.delegates.swap(new_delegates);
-            token
-        })
+        *guard = Some(new_list);
```

Comment:
The only remaining behavioral difference is that the old delegates aren't dropped outside the lock, as far as I can tell. That's what the `_lock_free_drop` does in the existing implementation for add/remove/clear. I'm not sure if there's a simple way to do that explicitly with `RwLock`.

Reply:
I think this will do
```Rust
/// Registers a delegate with the event object.
    pub fn add(&self, delegate: &T) -> Result<i64> {
        let mut _lock_free_drop = None;
        Ok({
            let mut guard = self.delegates.write().unwrap();
            let new_delegate = Delegate::new(delegate)?;
            let token = new_delegate.to_token();
            let new_iter = once(new_delegate);

            let new_list = if let Some(old_delegates) = guard.as_ref() {
                Arc::from_iter(old_delegates.iter().cloned().chain(new_iter))
            } else {
                Arc::from_iter(new_iter)
            };

            _lock_free_drop = guard.replace(new_list);
            token
        })
    }
```

Reply:
Looks good.

Reply:
I think using an explicit sequence like this:

```
drop(guard);
drop(old_list); // drop the old list _after_ releasing lock
```

would be more clear, and less likely to be accidentally changed during a refactor.